[MSE][GStreamer] Support markEndOfStream() before appendBuffer()

eocanha opened the pull request:
#### b1ff5913b21169adf6d94f74538d8ee90188abb6
<pre>
[MSE][GStreamer] Support markEndOfStream() before appendBuffer()
<a href="https://bugs.webkit.org/show_bug.cgi?id=278726">https://bugs.webkit.org/show_bug.cgi?id=278726</a>

Reviewed by Xabier Rodriguez-Calvar.

In the past, WebKit MSE couldn&apos;t complete playback when JavaScript
performs MediaSource.markEndOfStream() before appending any samples with
SourceBuffer.appendBuffer(). A previous patch added that support, but
still the error keeps happening under some circumstances downstream
(wpe-2.46, GStreamer 1.18) as well as upstream (where it was believed
not to happen in GStreamer 1.24).

This happens when the test case is exercised in a loop and the race
condition for an error occurs while errors are ignored, but the
GstMessage with the error takes a moment to propagate to the main thread
(when the errors are no longer ignored).

See: <a href="https://github.com/WebPlatformForEmbedded/WPEWebKit/issues/1558">https://github.com/WebPlatformForEmbedded/WPEWebKit/issues/1558</a>

This patch now applies the fix to any GStreamer version (because the
issue appears in all of them). When an EOS without buffers happens, the
pipeline goes to READY, and errors are ignored. The m_ignoreErrors flag
is kept enabled while the pipeline is being set to READY, but if any
error has been detected by the sync message handler, it stays enabled
until the last enqueued error is processed by the main thread. In any
case m_ignoreErrors is only changed from the main thread (in different
loop cycles in the worst case, but always from the main thread) and
doesn&apos;t need concurrent access protection.

As mentioned in the previous paragraph, a new sync error signal listener
has been added. This listener doesn&apos;t do any actual error processing,
just increases the m_queuedSyncErrors counter. The async message handler
will end up getting the error message anyway. That&apos;s where the message
will be ignored (if the m_ignoreErrors flag is set) and, ignored or not,
the m_queuedSyncErrors counter will be decremented. Note that this
counter is an Atomic, in order to support access from the (potentially)
non-main thread that handles the sync error signal, from the main thread
that ends up processing the error, and from the main thread that
processes setEosWithNoBuffers().

If an error has happened, an EOS in the pipeline is simulated by calling
didEnd(), the same method that would be called if such an EOS had
happened. This call usually triggers a pipeline teardown and the further
destruction of the player private. This may trigger a
gst_bus_set_flushing() call and cause the error message to never be
dispatched to the main thread. While this may mean that the
m_queuedSyncErrors counter is never decremented, this isn&apos;t a real
problem, since the player is being destructed anyway and nobody cares
about the count anymore.

* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::handleSyncErrorMessage): Detect errors as they happen, in the same thread where they were generated (often a non-main thread). The count of m_queuedSyncErrors is atomically increased to indicate the main thread that there are error messages pending to process. There&apos;s no need to forward the processing to the main thread, since the bus will eventually dispatch it asynchronously.
(WebCore::MediaPlayerPrivateGStreamer::handleMessage): If the last error is being processed while in a &quot;ignore errors&quot; state, reset the state to false. Also, exit early if we were in &quot;ignore errors&quot; state. In any case (early or standard exit), decrease the m_queuedSyncErrors count atomically.
(WebCore::MediaPlayerPrivateGStreamer::createGSTPlayBin): Process error messages synchronously with the handleSyncErrorMessage() handler.
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.h: didEnd() is now protected, so it can be called by the subclass. Added the m_queuedSyncErrors atomic counter to signal how many errors are waiting to be processed by the main thread.
* Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:
(WebCore::MediaPlayerPrivateGStreamerMSE::setEosWithNoBuffers): Ignore the errors on all GStreamer versions by declaring m_ignoreErrors as true. If pending error messages are enqueued, report EOS by calling didEnd() (m_ignoreErrors will be reset when the last error is processed). If no pending error messages appeared, just reset the m_ignoreErrors flag here.

Canonical link: <a href="https://commits.webkit.org/300605@main">https://commits.webkit.org/300605@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5e9fe3b56bec9d98202ccc8e36260b180c523246

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/123206 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/42920 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/33617 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/129913 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/75318 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/125083 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/43644 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/51515 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/93657 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/75318 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/126159 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/34785 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/110258 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/74287 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/33754 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/28415 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/73429 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/104498 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/28640 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/132629 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/50156 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/38180 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/102148 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/50532 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/106479 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/102006 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25932 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/47371 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/25582 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/46930 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/50011 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/55772 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/49481 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/52832 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/51160 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

eocanha commented:
Submitted a new version that doesn't use timeouts.

calvaris reviewed (changes requested):
There's a typo on the changelog, one wrongly written "couldn't"

eocanha commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.h:
```diff
@@ -442,6 +445,7 @@ class MediaPlayerPrivateGStreamer
     std::optional<GstVideoDecoderPlatform> m_videoDecoderPlatform;
     GstSeekFlags m_seekFlags;
     bool m_ignoreErrors { false };
+    Atomic<unsigned> m_queuedSyncErrors { 0 };
```
No. `m_ignoreErrors` is not in the critical section because it's only read/written by the main thread [here](https://github.com/WebKit/WebKit/pull/51211/files#diff-4376b854280205494dc9de0f08f985afd4d6584a7487c0a136adac2954c003cdR2077) and [here](https://github.com/WebKit/WebKit/pull/51211/files#diff-c1ec994b174b42858bb985196da2f4276385603280ca3772d8c267f8cd8cb7dcR542). The only variable that is used from multiple threads ([written from a (potentially) non-main thread](https://github.com/WebKit/WebKit/pull/51211/files#diff-4376b854280205494dc9de0f08f985afd4d6584a7487c0a136adac2954c003cdR1894) as well as [from the main thread](https://github.com/WebKit/WebKit/pull/51211/files#diff-4376b854280205494dc9de0f08f985afd4d6584a7487c0a136adac2954c003cdR1902) and read by the main thread [here](https://github.com/WebKit/WebKit/pull/51211/files#diff-4376b854280205494dc9de0f08f985afd4d6584a7487c0a136adac2954c003cdR2083)) and [here](https://github.com/WebKit/WebKit/pull/51211/files#diff-c1ec994b174b42858bb985196da2f4276385603280ca3772d8c267f8cd8cb7dcR550) is `m_queuedSyncErrors`. As it's a single variable, an `Atomic` is enough.

However, thanks to double-checking the code because of your comment, I've realized that I should be using +1 instead of -1 in the [`m_queuedSyncErrors.exchangeSub()` call](https://github.com/WebKit/WebKit/pull/51211/files#diff-4376b854280205494dc9de0f08f985afd4d6584a7487c0a136adac2954c003cdR1902) and that I can simplify the `m_ignoreErrors` logic [here](https://github.com/WebKit/WebKit/pull/51211/files#diff-c1ec994b174b42858bb985196da2f4276385603280ca3772d8c267f8cd8cb7dcR538) assuming that its value won't be externally changed throughout the method call. Thanks for that!

calvaris replied:
Ok, let's `ASSERT(isMainThread())` then just in case.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/3f0420e44b258ff2bf4fc901f1d5de634543ee34. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51211

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/2412bc75d169277549c4fe87dc902e07d7122b83. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51211

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/5e9fe3b56bec9d98202ccc8e36260b180c523246. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51211

eocanha committed:
[MSE][GStreamer] Support markEndOfStream() before appendBuffer()
https://bugs.webkit.org/show_bug.cgi?id=278726

Reviewed by Xabier Rodriguez-Calvar.

In the past, WebKit MSE couldn't complete playback when JavaScript
performs MediaSource.markEndOfStream() before appending any samples with
SourceBuffer.appendBuffer(). A previous patch added that support, but
still the error keeps happening under some circumstances downstream
(wpe-2.46, GStreamer 1.18) as well as upstream (where it was believed
not to happen in GStreamer 1.24).

This happens when the test case is exercised in a loop and the race
condition for an error occurs while errors are ignored, but the
GstMessage with the error takes a moment to propagate to the main thread
(when the errors are no longer ignored).

See: https://github.com/WebPlatformForEmbedded/WPEWebKit/issues/1558

This patch now applies the fix to any GStreamer version (because the
issue appears in all of them). When an EOS without buffers happens, the
pipeline goes to READY, and errors are ignored. The m_ignoreErrors flag
is kept enabled while the pipeline is being set to READY, but if any
error has been detected by the sync message handler, it stays enabled
until the last enqueued error is processed by the main thread. In any
case m_ignoreErrors is only changed from the main thread (in different
loop cycles in the worst case, but always from the main thread) and
doesn't need concurrent access protection.

As mentioned in the previous paragraph, a new sync error signal listener
has been added. This listener doesn't do any actual error processing,
just increases the m_queuedSyncErrors counter. The async message handler
will end up getting the error message anyway. That's where the message
will be ignored (if the m_ignoreErrors flag is set) and, ignored or not,
the m_queuedSyncErrors counter will be decremented. Note that this
counter is an Atomic, in order to support access from the (potentially)
non-main thread that handles the sync error signal, from the main thread
that ends up processing the error, and from the main thread that
processes setEosWithNoBuffers().

If an error has happened, an EOS in the pipeline is simulated by calling
didEnd(), the same method that would be called if such an EOS had
happened. This call usually triggers a pipeline teardown and the further
destruction of the player private. This may trigger a
gst_bus_set_flushing() call and cause the error message to never be
dispatched to the main thread. While this may mean that the
m_queuedSyncErrors counter is never decremented, this isn't a real
problem, since the player is being destructed anyway and nobody cares
about the count anymore.

* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::handleSyncErrorMessage): Detect errors as they happen, in the same thread where they were generated (often a non-main thread). The count of m_queuedSyncErrors is atomically increased to indicate the main thread that there are error messages pending to process. There's no need to forward the processing to the main thread, since the bus will eventually dispatch it asynchronously.
(WebCore::MediaPlayerPrivateGStreamer::handleMessage): If the last error is being processed while in a "ignore errors" state, reset the state to false. Also, exit early if we were in "ignore errors" state. In any case (early or standard exit), decrease the m_queuedSyncErrors count atomically.
(WebCore::MediaPlayerPrivateGStreamer::createGSTPlayBin): Process error messages synchronously with the handleSyncErrorMessage() handler.
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.h: didEnd() is now protected, so it can be called by the subclass. Added the m_queuedSyncErrors atomic counter to signal how many errors are waiting to be processed by the main thread.
* Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:
(WebCore::MediaPlayerPrivateGStreamerMSE::setEosWithNoBuffers): Ignore the errors on all GStreamer versions by declaring m_ignoreErrors as true. If pending error messages are enqueued, report EOS by calling didEnd() (m_ignoreErrors will be reset when the last error is processed). If no pending error messages appeared, just reset the m_ignoreErrors flag here.

Canonical link: https://commits.webkit.org/300605@main
